### ISSUE TITLE
coeio is broken after fork() / box.cfg {background = true}

### ISSUE THREAD

**Issue by rtsisyk** [tarantool/tarantool#517]
/etc/tarantool/instances.enabled/test.lua:

```
#!/usr/bin/env tarantool

local log = require('log')

box.cfg{}

log.info('before server')
require('socket').tcp_server('0.0.0.0', 22222, function(s) log.info('Connection: %s', s); s:write('hello\n'); s:close() end)
login.info('after server')
```

/var/log/tarantool/test.log:

```
2014-09-19 16:21:27.273 [21395] main/101/tarantoolctl C> version 1.6.3-344-g5482dc6
2014-09-19 16:21:27.273 [21395] main/101/tarantoolctl C> log level 5
2014-09-19 16:21:27.274 [21395] main/101/tarantoolctl I> mapping 1073741824 bytes for a shared arena...
2014-09-19 16:21:27.274 [21395] main/101/tarantoolctl I> start sophia recover
2014-09-19 16:21:27.274 [21396] main/101/spawner C> initialized
2014-09-19 16:21:27.275 [21395] main/101/tarantoolctl I> complete
2014-09-19 16:21:27.284 [21395] main/101/tarantoolctl I> initializing an empty data directory
2014-09-19 16:21:27.289 [21395] main/101/tarantoolctl I> creating `/var/lib/tarantool/test/00000000000000000000.snap.inprogress'
2014-09-19 16:21:27.289 [21395] main/101/tarantoolctl I> saving snapshot `/var/lib/tarantool/test/00000000000000000000.snap.inprogress'
2014-09-19 16:21:27.390 [21395] main/101/tarantoolctl I> done
2014-09-19 16:21:27.391 [21395] main/101/tarantoolctl I> ready to accept requests
2014-09-19 16:21:27.391 [21395] main/102/snap_daemon I> started
2014-09-19 16:21:27.391 [21395] main/101/test I> Run console at /var/run/tarantool/test.control
2014-09-19 16:21:27.391 [21395] main/101/test I> console: started on unix/:/var/run/tarantool/test.control
2014-09-19 16:21:27.391 [21395] main/101/test I> before server
```

```
[root@try instances.enabled]# netstat -a -n -p|grep 22222 <!-- nothing
```

```
[root@try instances.enabled]# socat UNIX-CONNECT:/var/run/tarantool/test.control STDIN
require('fiber').info()
---
- lua:
    fid: 103
    csw: 3
    backtrace:
    - '#0  0x48aec7 in fiber_yield_timeout(double)+103'
    - '#1  0x4b0aea in lbox_bsdsocket_iowait(lua_State*)+202'
    - '#2  0x4fea48 in lj_BC_FUNCC+52'
  snap_daemon:
    fid: 102
    csw: 1
    backtrace:
    - '#0  0x48aec7 in fiber_yield_timeout(double)+103'
    - '#1  0x48af19 in fiber_sleep(double)+9'
    - '#2  0x4ac3cf in lbox_fiber_sleep(lua_State*)+79'
    - '#3  0x4fea48 in lj_BC_FUNCC+52'
  test:
    fid: 101
    csw: 11
    backtrace:
    - '#0  0x48aec7 in fiber_yield_timeout(double)+103'
    - '#1  0x490a52 in coeio_custom+226'
    - '#2  0x4b075a in lbox_bsdsocket_getaddrinfo(lua_State*)+186'   <!--- HERE
    - '#3  0x4fea48 in lj_BC_FUNCC+52'
...
```

This code works fine without init scripts:
```
tarantool> require('socket').tcp_server('0.0.0.0', 22223, function(s) log.info('Connection: %s', s); s:write('hello\n'); s:close() end)
---
- s:
  - fd: 6
    name:
      host: 0.0.0.0
      family: AF_INET
      type: SOCK_STREAM
      protocol: tcp
      port: 22223
  stop: 'function: 0x40a65518'
...

tarantool> require('fiber').info()
---
- listen_fd=6:
    fid: 102
    csw: 1
    backtrace:
    - '#0  0x48aec7 in fiber_yield_timeout(double)+103'
    - '#1  0x4b0aea in lbox_bsdsocket_iowait(lua_State*)+202'
    - '#2  0x4fea48 in lj_BC_FUNCC+52'
  interactive:
    fid: 101
    csw: 6
    backtrace:
    - '#0  (nil) in ?'
...
```

**Comment by kostja** [comment]
A possible way to fix:
- *always* fork before initializing coeio. Set up pipes/redirection to redirect stdin/stdout to a child.
- if background option is set in box.cfg{}, exit from the parent process, which was attached to the terminal.
Alternatively, let the parent hang around so that it can work as a spawner/waitpid manager so that we can implement coio_*(fork/exec), popen in our API. We need fork/exec/popen for instance management, which is in turn necessary for automatic failover/replica reprovisioning and automatic testing.

It's best if this code is put into a library wiith a standard api in src/lib, and this library can be used by any fat process which needs to be able to daemonize/fork/exec and can't fork itself.

**Comment by mejedi** [comment]
Coeio is already [reinited](https://github.com/tarantool/tarantool/blob/master/src/main.cc#L373) on fork.

Reinit proc just literally re-initializes eio library, discarding the previous state altogether. In case there were any async tasks posted they won't complete, ever (hung fibers).

We can extend eio with a function to wait until async task queue drains and we can call it before fork.

In order for this bug to manifest itself, one has to spawn additional fibers calling coeio *before* doing fork/box.cfg{background=true}/. Since we are having a race condition here the failures won't be deterministic.

Whether it's worth fixing is questionable. Personally I believe it SHOULD be impossible to write a code exploiting this issue, even if the code is convoluted and it is highly unlikely to be encountered in practice.

**Note 1:** calling box.cfg{} again complains about setting option 'background' dynamically, hence it is probably safe to assume fork happens only once. No test for setting 'background' option dynamically, there are tests for other similar options (custom_proc_title, etc).

**Note 2:** initialising database engine spawns threads. The engine fails to initialise unless box.cfg{} was already called, hence the issue doesn't apply to these threads.

**Note 3:** initialising db before calling box.cfg{} fails with somewhat wordy error (due to traceback). It would be more friendly to just say 'Please call box.cfg{} first'. Maybe it makes sense to call box.cfg{} implicitly in this case?

**Comment by mejedi** [comment]
Original issue as reported in the very first message was about fork breaking fibers sheduling on FreeBSD (sample code never makes it through to the "after server" line).

This happens because libev REQUIRES certain actions to keep the event loop operational after fork. This is because kqueue has a super weird fork semantics and has to be recreated after fork.

The fix: https://github.com/tarantool/tarantool/commits/gh-517

**Comment by kostja** [comment]
* Nick Zavaritsky <notifications@github.com> [15/08/20 16:41]:
> Original issue as reported in the very first message was about fork breaking fibers sheduling on FreeBSD (sample code never makes it through to the "after server" line).
> 
> This happens because libev REQUIRES certain actions to keep the event loop operational after fork. This is because kqueue has a super weird fork semantics and has to be recreated after fork.
> 
> The fix: https://github.com/tarantool/tarantool/commits/gh-517

OK to push.


-- 
http://tarantool.org - a NoSQL database in a Lua script

**Comment by kostja** [comment]
Moving out to 1.7 since we decided to fix it by removing -b

**Comment by kostja** [comment]
We added re-initialization of coeio after fork.
The automatic restart feature is implemented in systemd.
Closing as won't fix.
